### ISSUE TITLE
Subtitle syncronization 

### ISSUE THREAD

**Issue by xavierfav** [Web-Multi-Media/HttpStreamingServer#36]
Currently we rely on [subliminal ](https://github.com/Diaoul/subliminal)to retrieve subtitles of the videos. Most of the times, the subtitles are already synchronized, but it happens that it's not the case. 

Solutions:

- One solution could be to include some re-sync shift time functionality in the frontend player. I think @worms19 already got something functional. However this means that each user has to re-sync the subtitles every time for each video. It would be possible to store the re-sync shift correction value set by one user in the backend database, so we could re-use the value for other users. Nonetheless, it can still be annoying for one person to have to set the values for each videos.

- We could try to download more candidate subtitles for the video (subliminal allows to do that easily), and allow the user to choose the right one. Again this information could be stored in database and only good subtitles could be kept in the backend server once verified.

- Maybe we could rely on automatic tools for synchronizing the subtitles. I found [this repo](https://github.com/oseiskar/autosubsync ) that could be interesting to test (it also gives references to other publicly available libraries).

**Comment by DerouineauNicolas** [comment]
- Regarding the first solution, I think the subtitles synchronization in the frontend would raise several issues. If we are adding a global offset to the subtitles, there would be some cases where the subtitles are synced at the beginning of the movie, and out of sync in the middle (therefore adding a global offset would not work). We would then need to handle offsets on sub parts (really tedious).

- I think we can try your second suggestions first (would be easier to implement). We would then have a dedicated API for subtitles (should improve the video list serializer performance).

- Really interesting library. I agree this is worth testing , but maybe only on user input ?  Like a tickbox saying "This sub is out of sync" which would then trigger backend processing. This would save us some precious processing time.

**Comment by xavierfav** [comment]
- I think there exist a speed control for the subtitles to solve these types of problems.

- Yes I guess it makes sense to try this solution first!

- That is exactly the idea. When a user reports/flags a video with wrong subtitles sync, we could launch this kind of re-sync processes.

---

So for now we decide to start by retrieving more subtitle files (2nd solution). Let's see if @yvan-chamoin is interested in that. In that case, @DerouineauNicolas, can you elaborate a bit on this point? What functions need to be changed for retrieving more subtitles?

**Comment by DerouineauNicolas** [comment]
The subtitles download function is defined [here]( (https://github.com/Web-Multi-Media/HttpStreamingServer/blob/master/backend/StreamServerApp/subtitles.py#L19)). It was adapted from the documentation of subliminal (https://subliminal.readthedocs.io/en/latest/user/usage.html).

The first step would to be replace the call to download_best_subtitles by individual listing, scoring and downloading  functions defined in the above subliminal documentation.

**Comment by xavierfav** [comment]
We had a look with @yvan-chamoin. It seems that subliminal does not always propose several subtitles. We tried with a video that gives problems in production and we got only one candidate subtitles that is not correctly synchronized. So I guess we should look into other solutions for having the subtitles synchronized. 

For now, @yvan-chamoin is going to optimize a bit the retrieval of the subtitles. It seems that we can easily perform a single request for the different subtitle languages. 

Moreover, this idea could also be extended to retrieve subtitles for several video in a single request, which could make our populate_db command much faster! We could also then create video by bulk in the database. We'll see later if we can work on that.
